### ISSUE TITLE
Unable to install rust on Raspberry Pi 3A+

### ISSUE THREAD

**Issue by shytikov** [rust-lang/rustup#2564]
Sorry for ditching standard template, but It's not really applicable here for me because I hit issue earlier on the process.

Regardless what I do I'm not able to install `rust` on my `Raspberry Pi 3A+` with 512MB of memory.

And even worse, installer leaves me with partially broken installation which I'm not able to cleanup afterwards.

Here is the log of `rustup` execution:

```
info: profile set to 'default'
info: default host triple is armv7-unknown-linux-gnueabihf
info: syncing channel updates for 'stable-armv7-unknown-linux-gnueabihf'
info: latest update on 2020-11-19, rust version 1.48.0 (7eac88abb 2020-11-16)
info: downloading component 'cargo'
  4.7 MiB /   4.7 MiB (100 %)   3.3 MiB/s in  2s ETA:  0s
info: downloading component 'clippy'
info: downloading component 'rust-std'
 18.2 MiB /  18.2 MiB (100 %)   3.7 MiB/s in  6s ETA:  0s
info: downloading component 'rustc'
 66.8 MiB /  66.8 MiB (100 %)   4.4 MiB/s in 19s ETA:  0s
info: downloading component 'rustfmt'
  3.1 MiB /   3.1 MiB (100 %)   3.1 MiB/s in  1s ETA:  0s
info: installing component 'cargo'
info: Defaulting to 329.6 MiB unpack ram
  4.7 MiB /   4.7 MiB (100 %)   4.0 MiB/s in  1s ETA:  0s
info: installing component 'clippy'
info: installing component 'rust-std'
 18.2 MiB /  18.2 MiB (100 %)   2.5 MiB/s in  9s ETA:  0s
info: installing component 'rustc'
 56.0 MiB /  66.8 MiB ( 84 %)   3.3 MiB/s in 30s ETA:  3sKilled
```

This issue is related to #2365, but since it was closed, creating a new one.

**Comment by kinnison** [comment]
Frankly this looks like it's running out of RAM installing the component which suggests that you're likely to hit similar problems compiling any non-trivial programs.  Either add swap, or cross-build from another more powerful system.

I am not sure what else to suggest.

**Comment by cztomsik** [comment]
Happens to me too, it worked fine before (1 year ago for sure), for what exactly does rustup needs so much memory?

**Comment by rbtcollins** [comment]
Try exporting RUSTUP_IO_THREADS=1 before running rustup.

The underlying problem is that rust has some very large files (single files > 200MB), and we need to untar the file and write it to disk, but because of interactions with other system constraints like Windows Defender and NFS servers we've settled on concurrent file writing across the board (it took several evolutions to get here, even though it doesn't look like an obvious place to aim); our next iteration is to introduce chunking for large files to deal with this issue, but none of our contributors have had time to write the fairly modest changes needed to do this.

**Comment by cztomsik** [comment]
It helped a bit, it failed first time but after restart it probably fit into memory and got through.

Oh, I had no idea, I thought the install process just downloads binaries and untars them somewhere (with system tar if possible). Ok, thanks a lot.

BTW: could you outline the work needed for handling big files on devices with limited memory?

**Comment by kinnison** [comment]
I am not certain of the details since I've not looked at that part of the codebase in a while, however in brief the work would involve detecting when a particularly large file is encountered in the incoming stream of files, and rather than reading it all in at once and passing it to an IO thread to be written out, operating on the main thread to do that particular file in a streaming fashion.

I'm sure @rbtcollins could do a better job of explaining how that might look, if he has some time.

**Comment by rbtcollins** [comment]
Fixed by #2707

**Comment by ixje** [comment]
For whomever finds this thread while trying to install on raspberry pi zero 2w, I had to `export RUSTUP_UNPACK_RAM=250000000` to get it to work
